small/ibuf.c

```diff
@@ -80,27 +80,40 @@ ibuf_reserve_slow(struct ibuf *ibuf, size_t size)
 	 */
 	if (size + used <= capacity) {
 		memmove(ibuf->buf, ibuf->rpos, used);
+		ibuf->rpos = ibuf->buf;
+		ibuf->wpos = ibuf->rpos + used;
+		return ibuf->wpos;
 	} else {
-		/* Use iobuf_readahead as allocation factor. */
 		size_t new_capacity = capacity * 2;
-		if (new_capacity < ibuf->start_capacity)
-			new_capacity = ibuf->start_capacity;
+		if (new_capacity < used + size)
+			new_capacity = used + size;
+		return ibuf_resize(ibuf, new_capacity);
+	}
+}
+
+void *
+ibuf_resize(struct ibuf *ibuf, size_t new_size)
+{
+	/* Ensure the existing data is not truncated. */
+	size_t used = ibuf_used(ibuf);
+	if (new_size < used)
+		new_size = used;
 
-		while (new_capacity < used + size)
-			new_capacity *= 2;
+	/* Use start_capacity as allocation factor. */
+	size_t new_capacity = ibuf->start_capacity;
+	while (new_capacity < new_size)
+		new_capacity *= 2;
```

Comment:
When we shrink the buffer, do we need to use the loop to compute the new capacity? Can't we just allocate the slab of the size closest to the current usage?

Reply:
Done.